src/Microsoft.Health.Dicom.Operations.Functions/DurableFunctionsCleanup.cs

```diff
@@ -0,0 +1,76 @@
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Operations.Functions.Indexing.Configuration;
+
+namespace Microsoft.Health.Dicom.Operations.Functions
+{
+    /// <summary>
+    /// Represents the Azure Functions that perform the cleanup of completed Durable Functions used for re-indexing.
+    /// </summary>
+    public class DurableFunctionsCleanup
+    {
+        private readonly ReindexErrorConfiguration _cleanupConfig;
+
+        // TODO: Leverages services for fulfilling the various activies
+        public DurableFunctionsCleanup(IOptions<ReindexErrorConfiguration> cleanupOptions)
+        {
+            EnsureArg.IsNotNull(cleanupOptions, nameof(cleanupOptions));
+            _cleanupConfig = cleanupOptions.Value;
+        }
+
+        // Currently an HTTP trigger, TODO: change to the appropriate trigger.
+        [FunctionName("DurableFunctionsCleanup")]
+        public async Task<String> Run(
+            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
+            [DurableClient] IDurableOrchestrationClient client,
+            ILogger log,
+            CancellationToken hostCancellationToken)
+        {
+            EnsureArg.IsNotNull(req, nameof(req));
+            EnsureArg.IsNotNull(client, nameof(client));
+            using var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(hostCancellationToken, req.HttpContext.RequestAborted);
+
+            log.LogInformation("Durable Functions cleanup started.");
+
+            EnsureArg.IsNotNull(client, nameof(client));
+
+            // Specify conditions for Function Queries.
+            OrchestrationStatusQueryCondition condition = new OrchestrationStatusQueryCondition
+            {
+                RuntimeStatus = _cleanupConfig.RuntimeStatuses,
+                CreatedTimeFrom = DateTime.MinValue,
+                CreatedTimeTo = DateTime.UtcNow.AddDays(-_cleanupConfig.MinimumAgeDays)
+            };
+
+            OrchestrationStatusQueryResult listOfOrchestrators =
+                await client.ListInstancesAsync(condition, cancellationTokenSource.Token);
+
+            EnsureArg.IsNotNull(listOfOrchestrators, nameof(listOfOrchestrators));
+
+            IEnumerable<String> nameList = listOfOrchestrators.DurableOrchestrationState.Select(i => i.Name);
+
+            // Loop through the orchestration instances and purge them.
+            foreach (var orch in listOfOrchestrators.DurableOrchestrationState)
```

Comment:
You actually need to use the resulting `OrchestrationStatusQueryResult` a bit more. If you inspect the [type](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.webjobs.extensions.durabletask.orchestrationstatusqueryresult?view=azure-dotnet), you'll see it has a property called `ContinuationToken` which is used to paginate the results. So if the Continuation is not `null`, you'll need to call `ListInstancesAsync` again with the token to get the next set of results.

Luckily, there are some new abstractions added to .NET, for operating on what we could consider to be an "asynchronous" enumeration (because we'd asynchronously fetch chunks of the elements we'd enumerate over):

```csharp
private async IAsyncEnumerable<DurableOrchestrationStatus> ListAllInstanceAsync(CancellationToken cancellationToken = default)
{
    OrchestrationStatusQueryCondition condition = new OrchestrationStatusQueryCondition  { ... };
    do
    {
        // TODO: Ensure the condition's ContinuationToken is updated based on the last batch
        OrchestrationStatusQueryResult batch = listOfOrchestrators = await client.ListInstancesAsync(condition, cancellationToken);

        // TODO: Yield the next batch of elements
    } while (batch.ContinuationToken != null);
}
```

Then, you can enumerate over this `IAsyncEnumerable<T>` using the new `await foreach` syntax.

Reply:
If this takes long enough, it may even warrant an orchestration!